storage/innobase/row/row0mysql.cc

```diff
@@ -228,24 +255,352 @@ row_mysql_read_true_varchar(
 
 	return(field + 1);
 }
+#define COLUMN_COMPRESS_PREFIX_MAX_LEN	5
+#define COLUMN_COMPRESS_HEADER_LEN	1
+#define COLUMN_COMPRESS_FLAG_MASK	(0x80)
+#define COLUMN_COMPRESS_FLAG		7	/*flag to mark if the column is compressed , bit 8*/
+#define COLUMN_COMPRESS_DATA_LEN_MASK	(0x60)
+#define COLUMN_COMPRESS_DATA_LEN	5	/*store bytes used to express the column length, bit 7,6*/
+#define COLUMN_COMPRESS_ALG_MASK	(0x1C)
+#define COLUMN_COMPRESS_ALG		2	/*store compression algorithm, default 0, bit 5,4,3*/
+#define COLUMN_COMPRESS_WRAP_MASK	(0x02)
+#define COLUMN_COMPRESS_WRAP		1	/*identify if adler32 is calculated, bit 2*/
+
+static void
+column_set_compress_header(
+	byte*	data,
+	bool	compressed,
+	ulint	lenlen,
+	uint	alg,
+	bool	wrap)
+{
+	data[0] &= 0x00;
+	data[0] |= (compressed << COLUMN_COMPRESS_FLAG);
+	data[0] |= (lenlen << COLUMN_COMPRESS_DATA_LEN);
+	data[0] |= (alg << COLUMN_COMPRESS_ALG);
+	data[0] |= (wrap << COLUMN_COMPRESS_WRAP);
+}
+
+static void
+column_get_compress_header(
+	const byte*	data,
+	bool*		compressed,
+	ulint*		lenlen,
+	uint*		alg,
+	bool*		wrap)
+{
+	*compressed = ((data[0] & COLUMN_COMPRESS_FLAG_MASK) >> COLUMN_COMPRESS_FLAG);
+	*lenlen = ((data[0] & COLUMN_COMPRESS_DATA_LEN_MASK) >> COLUMN_COMPRESS_DATA_LEN);
+	*alg = ((data[0] & COLUMN_COMPRESS_ALG_MASK) >> COLUMN_COMPRESS_ALG);
+	*wrap = ((data[0] & COLUMN_COMPRESS_WRAP_MASK) >> COLUMN_COMPRESS_WRAP);
+}
+
+static
+void*
+column_zip_zalloc(
+/*============*/
+	void*	opaque,	/*!< in/out: memory heap */
+	uInt	items,	/*!< in: number of items to allocate */
+	uInt	size)	/*!< in: size of an item in bytes */
+{
+	return(mem_heap_zalloc(static_cast<mem_heap_t*>(opaque), items * size));
+}
+
+/**********************************************************************//**
+Deallocate memory for zlib. */
+static
+void
+column_zip_free(
+/*==========*/
+	 void*	opaque MY_ATTRIBUTE((unused)),	/*!< in: memory heap */
+	 void*	address MY_ATTRIBUTE((unused)))	/*!< in: object to free */
+{
+}
+
+/**********************************************************************//**
+Configure the zlib allocator to use the given memory heap. */
+UNIV_INTERN
+void
+column_zip_set_alloc(
+/*===============*/
+	void*		stream,	/*!< in/out: zlib stream */
+	mem_heap_t*	heap)	/*!< in: memory heap to use */
+{
+	z_stream* strm = static_cast<z_stream*>(stream);
+
+	if (column_zip_mem_use_heap) {
+		strm->zalloc = column_zip_zalloc;
+		strm->zfree = column_zip_free;
+		strm->opaque = heap;
+	} else {
+		strm->zalloc = (alloc_func)0;
+		strm->zfree = (free_func)0;
+		strm->opaque = (voidpf)0;
+	}
+}
+
+/*******************************************************************//**
+Compress column data using zlib
+@return pointer to the compressed data */
+byte*
+row_compress_column(
+	const byte*	data,		/*!< in: data in mysql(uncompressed) format */
+	ulint*		len,		/*!< in: data length; out: length of compressed data*/
```

Comment:
Might not be practical, but all the "C array-like length" vars should be size_t here and elsewhere?